### DISCUSSION TITLE
Should Ruff check for use of incomaptible async APIs

### DISCUSSION THREAD

**Discussion by rafalkrupinski** [astral-sh/ruff#13159] [Ideas]
The use case is:
```py
async def a():
  import trio
  await trio.sleep(1)

async def b():
  import asyncio
  await asyncio.sleep(1)

async def c():
  await a()
  await b()
```

This will cause an exception since running any of those two `sleep` function requires the code to run in appropriate event loop, and it can't run in two event loops at the same time.

Would Ruff be the right project to check whether a function or any of its known callees makes call to competing async APIs?

Related [discussion in anyio](https://github.com/agronholm/anyio/discussions/543#discussioncomment-9982331).

**Comment by MichaReiser** [comment]
I guess that depends. Ruff focuses on linting your project files but I suspect you want to check that no method (including third-party code) mixes `trio` and `asyncio`. 

Either way. Ruff isn't powerful enough today to do this kind of analysis because the analysis is limited to a single file. We're working on the red knot to support multi-file analysis that would also allow us to follow call expressions to functions defined in other files (at least for as long as the callee isn't dynamic).

**Reply by rafalkrupinski** [reply]
yes, including indirect calls via libraries. This is invalid:
```
a()
  b()
    asyncio call()
  c()
    trio call()
  ```
I'm not sure which tool could implement this. Mypy checks cross-file, but it focuses on typing. This problem is similar, each function would need to be marked as asyncio/trio/curio specific or async agnostic, and callees of each function would need to be checked for mis-matches.
